diff --git a/T1-CPU/ifu/sparc_ifu_swl.v b/T1-CPU/ifu/sparc_ifu_swl.v
index 331b79d..7b4c242 100644
--- a/T1-CPU/ifu/sparc_ifu_swl.v
+++ b/T1-CPU/ifu/sparc_ifu_swl.v
@@ -198,6 +198,7 @@ module sparc_ifu_swl(/*AUTOARG*/
    // local signals
    wire [9:0]    count_nxt,
 	               count;
+   wire [3:0]    fast_boot_nextthr_bf_raw;
    wire          proc0;
    wire          start_on_rst;
 
@@ -1129,9 +1130,10 @@ module sparc_ifu_swl(/*AUTOARG*/
    assign use_spec = ~(rdy[3] | rdy[2] | rdy[1] | rdy[0]);
 
    assign sched_reset = dtu_reset | ~gdbginit_l;
+   assign dtu_fcl_nextthr_bf[3:0] = start_on_rst ? 4'b0001 : fast_boot_nextthr_bf_raw[3:0];
    // schedule ready threads
    sparc_ifu_lru4 thr_sched(// Outputs
-			                      .grant_vec	(dtu_fcl_nextthr_bf[3:0]),
+			                      .grant_vec	(fast_boot_nextthr_bf_raw[3:0]),
 			                      .so		(so),
 			                      // Inputs
 			                      .clk	(clk),
