diff --git a/T1-CPU/lsu/lsu.v b/T1-CPU/lsu/lsu.v
--- a/T1-CPU/lsu/lsu.v
+++ b/T1-CPU/lsu/lsu.v
@@ -614,6 +614,8 @@
 wire                    lsu_dtag_wrreq_x_e;     // From dctl of lsu_dctl.v
 wire                    lsu_dtagv_wr_vld_e;     // From dctl of lsu_dctl.v
 wire                    lsu_dtlb_addr_mask_l_e; // From dctl of lsu_dctl.v
+wire                    fast_boot_dtlb_bypass_e;
+wire                    fast_boot_lsu_dtlb_bypass_e;
 wire                    lsu_dtlb_bypass_e;      // From dctl of lsu_dctl.v
 wire [2:0]              lsu_dtlb_cam_pid_e;     // From dctldp of lsu_dctldp.v
 wire                    lsu_dtlb_data_rd_e;     // From dctl of lsu_dctl.v
@@ -1014,6 +1018,10 @@
 output [2:0]		lsu_ffu_bld_cnt_w ;
    
 wire [47:0]  lsu_local_ldxa_data_g;
+assign fast_boot_dtlb_bypass_e =
+  ~ifu_lsu_alt_space_d &
+  (exu_lsu_ldst_va_e[47:18] == 30'b0);
+assign fast_boot_lsu_dtlb_bypass_e = lsu_dtlb_bypass_e | fast_boot_dtlb_bypass_e;
 wire [43:0]  lsu_iobrdge_rd_data;	
 wire [79:0]  stb_rdata_ramd;
 wire [75:64]  stb_wdata_ramd_b75_b64;
@@ -2068,7 +2074,7 @@
                  .lsu_alt_space_m       (lsu_alt_space_m),
                  .tlb_cam_hit           (tlb_cam_hit),
                  .ifu_lsu_ld_inst_e     (ifu_lsu_ld_inst_e),
-                 .lsu_dtlb_bypass_e     (lsu_dtlb_bypass_e),
+                 .lsu_dtlb_bypass_e     (fast_boot_lsu_dtlb_bypass_e),
                  .cache_way_hit         (cache_way_hit[3:0]));
 
 /*
@@ -2583,7 +2589,7 @@
                .lsu_exu_dfill_vld_w2    (lsu_exu_dfill_vld_w2),
                .lsu_ffu_ld_vld          (lsu_ffu_ld_vld),
                .lsu_ld_miss_wb          (lsu_ld_miss_wb),
-               .lsu_dtlb_bypass_e       (lsu_dtlb_bypass_e),
+               .lsu_dtlb_bypass_e     (fast_boot_lsu_dtlb_bypass_e),
                .ld_pcx_pkt_g            (ld_pcx_pkt_g[`LMQ_WIDTH-1:40]),
                .tlb_ldst_cam_vld        (tlb_ldst_cam_vld),
                .ldxa_internal           (ldxa_internal),
@@ -3040,7 +3046,7 @@
                .lsu_exu_dfill_vld_w2    (lsu_exu_dfill_vld_w2),
                .lsu_ffu_ld_vld          (lsu_ffu_ld_vld),
                .lsu_ld_miss_wb          (lsu_ld_miss_wb),
-               .lsu_dtlb_bypass_e       (lsu_dtlb_bypass_e),
+               .lsu_dtlb_bypass_e     (fast_boot_lsu_dtlb_bypass_e),
                .ld_pcx_pkt_g            (ld_pcx_pkt_g[`LMQ_WIDTH-1:40]),
                .tlb_ldst_cam_vld        (tlb_ldst_cam_vld),
                .ldxa_internal           (ldxa_internal),
@@ -3549,7 +3555,7 @@
                  .rst_soft_l            (lsu_dtlb_invalid_all_l_m), 
                  .hold              	(sehold),                      
                  .tlb_addr_mask_l       (lsu_dtlb_addr_mask_l_e), 
-                 .tlb_bypass            (lsu_dtlb_bypass_e),     
+                 .tlb_bypass            (fast_boot_lsu_dtlb_bypass_e),     
                  .tlb_bypass_va         (exu_lsu_ldst_va_e[12:10]), 
                  .tlb_cam_pid           (lsu_dtlb_cam_pid_e[2:0]), 
                  //.tlb_cam_real          (lsu_dtlb_cam_real_e),   
@@ -3615,7 +3621,7 @@
                 .cache_ptag_w3          ({dtag_rdata_w3_m[28:0], lsu_ldst_va_m[10]}), // Templated
                 .cache_set_vld          (dva_vld_m[3:0]),        // Templated
                 .tlb_bypass_va          (exu_lsu_ldst_va_e[12:10]), // Templated
-                .tlb_bypass             (lsu_dtlb_bypass_e),     // Templated
+                .tlb_bypass             (fast_boot_lsu_dtlb_bypass_e),     // Templated
                 .se                     (se),
                 .hold                   (sehold),                // Templated
                 .adj                    (lsu_dtlb_mrgn[7:0]),    // Templated
@@ -4098,7 +4104,7 @@
                        .stb_state_rmo   (stb0_state_rmo[7:0]),   // Templated
                        .stb_alt_sel     (lsu_blk_st_m),          // Templated
                        .stb_alt_addr    (lsu_blkst_pgnum_m[39:37]), // Templated
-                       .lsu_dtlb_bypass_e(lsu_dtlb_bypass_e),
+                       .lsu_dtlb_bypass_e(fast_boot_lsu_dtlb_bypass_e),
                        .tlb_cam_hit     (tlb_cam_hit),
                        .st_dtlb_perr_g  (lsu_st_dtlb_perr_g[0]), // Templated
                        .lsu_outstanding_rmo_st_max(lsu_outstanding_rmo_st_max[0])); // Templated
@@ -4463,7 +4469,7 @@
                        .stb_state_rmo   (stb1_state_rmo[7:0]),   // Templated
                        .stb_alt_sel     (lsu_blk_st_m),          // Templated
                        .stb_alt_addr    (lsu_blkst_pgnum_m[39:37]), // Templated
-                       .lsu_dtlb_bypass_e(lsu_dtlb_bypass_e),
+                       .lsu_dtlb_bypass_e(fast_boot_lsu_dtlb_bypass_e),
                        .tlb_cam_hit     (tlb_cam_hit),
                        .st_dtlb_perr_g  (lsu_st_dtlb_perr_g[1]), // Templated
                        .lsu_outstanding_rmo_st_max(lsu_outstanding_rmo_st_max[1])); // Templated
@@ -4553,7 +4559,7 @@
                        .stb_state_rmo   (stb2_state_rmo[7:0]),   // Templated
                        .stb_alt_sel     (lsu_blk_st_m),          // Templated
                        .stb_alt_addr    (lsu_blkst_pgnum_m[39:37]), // Templated
-                       .lsu_dtlb_bypass_e(lsu_dtlb_bypass_e),
+                       .lsu_dtlb_bypass_e(fast_boot_lsu_dtlb_bypass_e),
                        .tlb_cam_hit     (tlb_cam_hit),
                        .st_dtlb_perr_g  (lsu_st_dtlb_perr_g[2]), // Templated
                        .lsu_outstanding_rmo_st_max(lsu_outstanding_rmo_st_max[2])); // Templated
@@ -4643,7 +4649,7 @@
                        .stb_state_rmo   (stb3_state_rmo[7:0]),   // Templated
                        .stb_alt_sel     (lsu_blk_st_m),          // Templated
                        .stb_alt_addr    (lsu_blkst_pgnum_m[39:37]), // Templated
-                       .lsu_dtlb_bypass_e(lsu_dtlb_bypass_e),
+                       .lsu_dtlb_bypass_e(fast_boot_lsu_dtlb_bypass_e),
                        .tlb_cam_hit     (tlb_cam_hit),
                        .st_dtlb_perr_g  (lsu_st_dtlb_perr_g[3]), // Templated
                        .lsu_outstanding_rmo_st_max(lsu_outstanding_rmo_st_max[3])); // Templated
