diff --git a/T1-FPU/fpu_rptr_groups.v b/T1-FPU/fpu_rptr_groups.v
--- a/T1-FPU/fpu_rptr_groups.v
+++ b/T1-FPU/fpu_rptr_groups.v
@@ -489,10 +489,17 @@
 
 	// buffers for se (scan enable driven from test_stub_scan)
 	fpu_bufrpt_grp4 i_se_buf1 (
+`ifdef NO_SCAN
+		.in ({se,
+			se,
+			1'b0,
+			1'b0}),
+`else
 		.in ({se,
 			se,
 			so_unbuf,
 			1'b0}),
+`endif
 		.out ({se_add_buf1,
 			se_mul64_buf1,
 			so_buf1,
