diff --git a/ao486/memory/icache.v b/ao486/memory/icache.v
--- a/ao486/memory/icache.v
+++ b/ao486/memory/icache.v
@@ -1,224 +1,119 @@
-/*
- * Copyright (c) 2014, Aleksander Osman
- * All rights reserved.
- * 
- * Redistribution and use in source and binary forms, with or without
- * modification, are permitted provided that the following conditions are met:
- * 
- * * Redistributions of source code must retain the above copyright notice, this
- *   list of conditions and the following disclaimer.
- * 
- * * Redistributions in binary form must reproduce the above copyright notice,
- *   this list of conditions and the following disclaimer in the documentation
- *   and/or other materials provided with the distribution.
- * 
- * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
- * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
- * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
- * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
- * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
- * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
- * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
- * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
- * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
- * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
- */
-
-`include "defines.v"
-
 module icache(
-    input           clk,
-    input           rst_n,
-    
-    input           cache_disable,
-
-    //RESP:
-    input           pr_reset,
-    
-    input [31:0]    prefetch_address,
-    input [31:0]    delivered_eip,
-    output reg      reset_prefetch = 1'd0,
-    //END
-    
-    //RESP:
-    input           icacheread_do,
-    input   [31:0]  icacheread_address,
-    input   [4:0]   icacheread_length, // takes into account: page size and cs segment limit
-    //END
-    
-    //REQ:
-    output              readcode_do,
-    input               readcode_done,
-    
-    output      [31:0]  readcode_address,
-    input       [31:0]  readcode_partial,
-    //END
-    
-    //REQ:
-    output              prefetchfifo_write_do,
-    output  [35:0]      prefetchfifo_write_data,
-    //END
-    
-    //REQ:
-    output              prefetched_do,
-    output [4:0]        prefetched_length,
-    //END
-    
-    input   [27:2]      snoop_addr,
-    input   [31:0]      snoop_data,
-    input    [3:0]      snoop_be,
-    input               snoop_we
+  input         clk,
+                rst_n,
+                cache_disable,
+                pr_reset,
+  input  [31:0] prefetch_address,
+                delivered_eip,
+  input         icacheread_do,
+  input  [31:0] icacheread_address,
+  input  [4:0]  icacheread_length,
+  input         readcode_done,
+  input  [31:0] readcode_partial,
+  input  [25:0] snoop_addr,
+  input  [31:0] snoop_data,
+  input  [3:0]  snoop_be,
+  input         snoop_we,
+  output        reset_prefetch,
+                readcode_do,
+  output [31:0] readcode_address,
+  output        prefetchfifo_write_do,
+  output [35:0] prefetchfifo_write_data,
+  output        prefetched_do,
+  output [4:0]  prefetched_length
 );
 
-//------------------------------------------------------------------------------
-
-localparam STATE_IDLE = 1'd0;
-localparam STATE_READ = 1'd1;
-
-reg          state;
-reg [4:0]    length;
-reg [11:0]   partial_length;
-reg          reset_waiting;
-             
-wire [4:0]   partial_length_current;
-
-wire [11:0]  length_burst;
-
-wire         readcode_cache_do;
-wire [31:0]  readcode_cache_address;
-wire         readcode_cache_valid;
-wire         readcode_cache_done;
-wire [31:0]  readcode_cache_data;
-
-reg          prefetch_checknext;
-reg  [31:0]  prefetch_checkaddr;
-reg  [31:0]  min_check;
-reg  [31:0]  max_check;
-reg   [1:0]  reset_prefetch_count = 2'd0;
-
-
-//------------------------------------------------------------------------------
-
-wire reset_combined = reset_prefetch | pr_reset;
-
-always @(posedge clk) begin
-   prefetch_checknext <= 1'b0;
-   prefetch_checkaddr <= { 4'd0, snoop_addr, 2'd0 };
-   min_check          <= delivered_eip;
-   max_check          <= prefetch_address + 5'd20; // cache read burst is 16 bytes, so we need to look a bit further, additional + 4 because of 1 cycle delay.
-   
-   if (snoop_we) prefetch_checknext <= 1'b1;
-   
-   if (prefetch_checknext && prefetch_checkaddr >= min_check && prefetch_checkaddr <= max_check) begin
-      reset_prefetch       <= 1'b1;
-      reset_prefetch_count <= 2'd2;
-   end
-   
-   if (reset_prefetch_count > 2'd0) begin
-      reset_prefetch_count <= reset_prefetch_count - 1'd1;
-      if (reset_prefetch_count == 2'd1) reset_prefetch <= 1'd0;
-   end
-   
-end
-
-//------------------------------------------------------------------------------
-
-//MIN(partial_length, length_saved)
-assign partial_length_current =
-    ({ 2'b0, partial_length[2:0] } > length)? length : { 2'b0, partial_length[2:0] };
-    
-//------------------------------------------------------------------------------
-
-always @(posedge clk) begin
-    if(rst_n == 1'b0)                                 reset_waiting <= `FALSE;
-    else if(reset_combined && state != STATE_IDLE)    reset_waiting <= `TRUE;
-    else if(state == STATE_IDLE)                      reset_waiting <= `FALSE;
-end
-
-//------------------------------------------------------------------------------
-
-assign length_burst =
-    (icacheread_address[1:0] == 2'd0)? { 3'd4, 3'd4, 3'd4, 3'd4 } :
-    (icacheread_address[1:0] == 2'd1)? { 3'd4, 3'd4, 3'd4, 3'd3 } :
-    (icacheread_address[1:0] == 2'd2)? { 3'd4, 3'd4, 3'd4, 3'd2 } :
-                                       { 3'd4, 3'd4, 3'd4, 3'd1 };
-                            
-assign prefetchfifo_write_data =
-    (partial_length[2:0] == 3'd1)?   {                  4'd1 ,              24'd0, readcode_cache_data[31:24] } :
-    (partial_length[2:0] == 3'd2)?   { (length > 5'd2)? 4'd2 : length[3:0], 16'd0, readcode_cache_data[31:16] } :
-    (partial_length[2:0] == 3'd3)?   { (length > 5'd3)? 4'd3 : length[3:0],  8'd0, readcode_cache_data[31:8] } :
-                                     { (length > 5'd4)? 4'd4 : length[3:0],        readcode_cache_data[31:0] };
-
-//------------------------------------------------------------------------------
-
-l1_icache l1_icache_inst(
-   
-    .CLK             (clk),
-    .RESET           (~rst_n),
-    .pr_reset        (reset_combined),
-	 
-    .DISABLE         (cache_disable),
-
-    .CPU_REQ         (readcode_cache_do),
-    .CPU_ADDR        (readcode_cache_address),
-    .CPU_VALID       (readcode_cache_valid),
-    .CPU_DONE        (readcode_cache_done),
-    .CPU_DATA        (readcode_cache_data),
-    
-    .MEM_REQ         (readcode_do),
-    .MEM_ADDR        (readcode_address),
-    .MEM_DONE        (readcode_done),
-    .MEM_DATA        (readcode_partial),
-    
-    .snoop_addr      (snoop_addr),
-    .snoop_data      (snoop_data),
-    .snoop_be        (snoop_be),
-    .snoop_we        (snoop_we)
-);
-
-assign readcode_cache_do =
-   (~rst_n) ? (`FALSE) :
-   (state == STATE_IDLE && ~(reset_combined) && icacheread_do && icacheread_length > 5'd0) ? (`TRUE) :
-   `FALSE;
-   
-assign readcode_cache_address = { icacheread_address[31:2], 2'd0 };
-   
-assign prefetchfifo_write_do =
-   (~rst_n) ? (`FALSE) :
-   (state == STATE_READ && reset_combined == `FALSE && reset_waiting == `FALSE && readcode_cache_valid) ? (`TRUE) :
-   `FALSE;
-   
-assign prefetched_length       = partial_length_current;
-
-assign prefetched_do =
-   (~rst_n) ? (`FALSE) :
-   (state == STATE_READ && reset_combined == `FALSE && reset_waiting == `FALSE && readcode_cache_valid) ? (`TRUE) :
-   `FALSE;
-   
-always @(posedge clk) begin
-   if(rst_n == 1'b0) begin
-      state          <= STATE_IDLE;
-      length         <= 5'b0;
-      partial_length <= 12'b0;
-   end
-   else begin
-      if(state == STATE_IDLE && ~(reset_combined) && icacheread_do && icacheread_length > 5'd0) begin
-         state          <= STATE_READ;
-         partial_length <= length_burst;
-         length         <= icacheread_length;
-      end
-      else if (state == STATE_READ) begin
-         if(reset_combined == `FALSE && reset_waiting == `FALSE) begin
-            if(readcode_cache_valid) begin
-               if(partial_length[2:0] > 3'd0 && length > 5'd0) begin
-                  length         <= length - partial_length_current;
-                  partial_length <= { 3'd0, partial_length[11:3] }; 
-               end
-            end
-         end
-         if(readcode_cache_done) state <= STATE_IDLE;
-      end
-   end
-end     
-
+  reg  [3:0]  rt_tmp_12_4;
+  reg  [11:0] rt_tmp_10_12;
+  reg  [4:0]  rt_tmp_9_5;
+  reg         rt_tmp_1_1;
+  reg  [31:0] rt_tmp_2_32;
+  reg  [31:0] rt_tmp_3_32;
+  reg  [31:0] rt_tmp_4_32;
+  reg         rt_tmp_5_1;
+  reg  [1:0]  rt_tmp_6_2;
+  wire        _GEN = rt_tmp_5_1 | pr_reset;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1135:17, :1145:18
+  reg         rt_tmp_7_1;
+  wire [4:0]  _GEN_0 = {2'h0, rt_tmp_10_12[2:0]};	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1167:18, :1168:18, :1255:19
+  wire [4:0]  _GEN_1 = _GEN_0 > rt_tmp_9_5 ? rt_tmp_9_5 : _GEN_0;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1168:18, :1169:18, :1170:18, :1224:17
+  wire        _GEN_2 = readcode_done & rt_tmp_12_4 < 4'h4 & (|_GEN_1) & ~pr_reset;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1165:18, :1166:18, :1170:18, :1172:18, :1173:18, :1175:18, :1176:18, :1177:18, :1286:18
+  reg         rt_tmp_8_1;
+  reg         rt_tmp_11_1;
+  always_ff @(posedge clk) begin
+    automatic logic _GEN_3 = rt_tmp_6_2 == 2'h0;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1120:18, :1143:17
+    automatic logic _GEN_4 = rt_tmp_6_2 == 2'h1;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1123:18, :1124:18, :1143:17
+    automatic logic _GEN_5 =
+      rt_tmp_1_1 & rt_tmp_2_32 >= rt_tmp_3_32 & rt_tmp_2_32 <= rt_tmp_4_32;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1111:17, :1115:18, :1116:18, :1119:18, :1126:18, :1127:18, :1128:18, :1129:18
+    automatic logic _GEN_6 = rst_n & _GEN & rt_tmp_8_1;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1145:18, :1146:18, :1147:18, :1191:17
+    automatic logic _GEN_7 = ~rt_tmp_8_1 & ~_GEN & icacheread_do & (|icacheread_length);	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1145:18, :1149:18, :1153:18, :1154:18, :1155:18, :1157:18, :1158:18, :1191:17
+    automatic logic _GEN_8 = ~_GEN_7 & rst_n & ~rt_tmp_8_1;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1149:18, :1154:18, :1155:18, :1158:18, :1159:18, :1160:18, :1161:18, :1191:17
+    automatic logic _GEN_9 = rst_n & _GEN_7;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1154:18, :1155:18, :1158:18, :1163:18
+    automatic logic _GEN_10 = ~rst_n | _GEN_9;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1144:18, :1163:18, :1164:18
+    automatic logic _GEN_11 = _GEN_2 & rt_tmp_12_4 == 4'h3 | pr_reset;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1173:18, :1176:18, :1177:18, :1178:18, :1179:18, :1180:18, :1181:18, :1286:18
+    automatic logic _GEN_12 = ~_GEN & ~rt_tmp_7_1;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1145:18, :1152:17, :1153:18, :1193:18, :1194:18
+    automatic logic _GEN_13 = _GEN_12 & ~_GEN_2;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1173:18, :1176:18, :1177:18, :1194:18, :1195:18, :1196:18
+    automatic logic _GEN_14 =
+      _GEN_2 & _GEN_12 & ~((|(rt_tmp_10_12[2:0])) & (|rt_tmp_9_5));	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1167:18, :1173:18, :1176:18, :1177:18, :1194:18, :1199:18, :1202:18, :1203:18, :1204:18, :1205:19, :1206:19, :1224:17, :1255:19
+    automatic logic _GEN_15 =
+      ~(~_GEN_8 & (_GEN_10 | _GEN_12 & ~_GEN_13 & ~_GEN_14)) | _GEN_8;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1160:18, :1161:18, :1162:18, :1164:18, :1194:18, :1196:18, :1197:18, :1198:18, :1199:18, :1206:19, :1207:19, :1208:19, :1209:19, :1210:19, :1211:19, :1212:19
+    automatic logic _GEN_16 = ~_GEN_12 | _GEN_13 | _GEN_14;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1194:18, :1196:18, :1199:18, :1206:19, :1213:19, :1214:19, :1215:19
+    automatic logic _GEN_17 = ~rst_n | pr_reset;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1144:18, :1258:19
+    automatic logic _GEN_18 = rt_tmp_12_4 == 4'h7;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1260:19, :1261:19, :1286:18
+    rt_tmp_1_1 <= snoop_we;
+    rt_tmp_2_32 <= {4'h0, snoop_addr, 2'h0};	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1112:18, :1114:19, :1115:18
+    rt_tmp_3_32 <= delivered_eip;
+    rt_tmp_4_32 <= prefetch_address + 32'h14;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1117:19, :1118:19, :1119:18
+    rt_tmp_5_1 <= ~_GEN_3 & _GEN_4 | _GEN_5 ? (_GEN_3 | ~_GEN_4) & _GEN_5 : rt_tmp_5_1;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1120:18, :1122:18, :1124:18, :1125:18, :1127:18, :1129:18, :1130:18, :1131:18, :1132:18, :1133:18, :1134:18, :1135:17
+    rt_tmp_6_2 <=
+      ~_GEN_3 | _GEN_5
+        ? (_GEN_3 ? (_GEN_5 ? 2'h2 : rt_tmp_6_2) : rt_tmp_6_2 - 2'h1)
+        : rt_tmp_6_2;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1120:18, :1122:18, :1127:18, :1129:18, :1136:18, :1137:18, :1138:18, :1140:18, :1141:18, :1142:18, :1143:17
+    rt_tmp_7_1 <= ~rst_n | _GEN_6 | ~rt_tmp_8_1 ? _GEN_6 : rt_tmp_7_1;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1144:18, :1146:18, :1147:18, :1148:18, :1149:18, :1150:18, :1151:18, :1152:17, :1191:17
+    rt_tmp_8_1 <=
+      ~(~_GEN_8 & (_GEN_10 | _GEN_11)) | _GEN_8
+        ? rt_tmp_8_1
+        : rst_n & (_GEN_9 | ~_GEN_11 & rt_tmp_8_1);	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1160:18, :1161:18, :1162:18, :1163:18, :1164:18, :1181:18, :1182:18, :1183:18, :1184:18, :1185:18, :1186:18, :1187:18, :1188:18, :1189:18, :1190:18, :1191:17
+    rt_tmp_9_5 <=
+      _GEN_15
+        ? rt_tmp_9_5
+        : ~rst_n
+            ? 5'h0
+            : _GEN_9 ? icacheread_length : _GEN_16 ? rt_tmp_9_5 : rt_tmp_9_5 - _GEN_1;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1144:18, :1156:18, :1163:18, :1170:18, :1212:19, :1214:19, :1215:19, :1219:19, :1220:19, :1221:19, :1222:19, :1223:19, :1224:17
+    rt_tmp_10_12 <=
+      _GEN_15
+        ? rt_tmp_10_12
+        : ~rst_n
+            ? 12'h0
+            : _GEN_9
+                ? (~(icacheread_address[1])
+                     ? (icacheread_address[1:0] == 2'h0 | icacheread_address[1:0] != 2'h1
+                          ? 12'h924
+                          : 12'h923)
+                     : icacheread_address[1:0] != 2'h3
+                         ? (icacheread_address[1:0] == 2'h2 ? 12'h922 : 12'h924)
+                         : (&(icacheread_address[1:0])) ? 12'h921 : 12'h924)
+                : _GEN_16 ? rt_tmp_10_12 : {3'h0, rt_tmp_10_12[11:3]};	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1123:18, :1137:18, :1139:18, :1144:18, :1163:18, :1201:18, :1212:19, :1214:19, :1215:19, :1225:20, :1226:19, :1228:19, :1230:19, :1231:20, :1233:19, :1234:20, :1236:20, :1238:19, :1240:19, :1241:20, :1242:20, :1244:19, :1245:20, :1246:20, :1247:20, :1248:20, :1249:19, :1250:20, :1251:20, :1252:20, :1253:20, :1254:20, :1255:19
+    rt_tmp_11_1 <= ~_GEN_17 & readcode_done & (~rt_tmp_11_1 | ~_GEN_18);	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1258:19, :1261:19, :1264:19, :1266:19, :1268:19, :1269:19, :1270:18
+    rt_tmp_12_4 <=
+      _GEN_17 | ~readcode_done
+        ? 4'h0
+        : rt_tmp_11_1 ? (_GEN_18 ? 4'h0 : rt_tmp_12_4 + 4'h1) : 4'h1;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1112:18, :1258:19, :1261:19, :1270:18, :1278:19, :1279:19, :1280:19, :1282:19, :1285:19, :1286:18
+  end // always_ff @(posedge)
+  wire        _GEN_19 = rst_n & rt_tmp_8_1 & ~_GEN & ~rt_tmp_7_1 & _GEN_2;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1145:18, :1152:17, :1153:18, :1173:18, :1176:18, :1177:18, :1191:17, :1193:18, :1297:19, :1298:19, :1300:19, :1301:19
+  assign reset_prefetch = rt_tmp_5_1;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1135:17, :1334:3
+  assign readcode_do = rst_n & ~rt_tmp_8_1 & ~_GEN & icacheread_do & (|icacheread_length);	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1145:18, :1149:18, :1153:18, :1157:18, :1191:17, :1288:19, :1291:19, :1292:19, :1294:19, :1334:3
+  assign readcode_address = {icacheread_address[31:2], 2'h0};	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1295:20, :1296:20, :1334:3
+  assign prefetchfifo_write_do = _GEN_19;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1297:19, :1298:19, :1300:19, :1301:19, :1334:3
+  assign prefetchfifo_write_data =
+    rt_tmp_10_12[2:0] == 3'h1
+      ? {28'h1000000, readcode_partial[31:24]}
+      : rt_tmp_10_12[2:0] == 3'h2
+          ? {rt_tmp_9_5 > 5'h2 ? 4'h2 : rt_tmp_9_5[3:0], 16'h0, readcode_partial[31:16]}
+          : rt_tmp_10_12[2:0] == 3'h3
+              ? {(|(rt_tmp_9_5[4:2])) ? 4'h3 : rt_tmp_9_5[3:0],
+                 8'h0,
+                 readcode_partial[31:8]}
+              : {rt_tmp_9_5 > 5'h4 ? 4'h4 : rt_tmp_9_5[3:0], readcode_partial};	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1165:18, :1167:18, :1178:18, :1224:17, :1255:19, :1302:19, :1303:19, :1304:20, :1305:19, :1306:20, :1307:19, :1308:19, :1309:19, :1310:19, :1311:19, :1312:19, :1313:19, :1314:20, :1315:20, :1316:20, :1317:19, :1318:19, :1319:19, :1320:19, :1322:19, :1323:19, :1324:20, :1325:20, :1326:19, :1327:19, :1329:19, :1330:20, :1331:20, :1332:20, :1333:20, :1334:3
+  assign prefetched_do = _GEN_19;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1297:19, :1298:19, :1300:19, :1301:19, :1334:3
+  assign prefetched_length = _GEN_1;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:1170:18, :1334:3
 endmodule

