diff --git a/ao486/pipeline/fetch.v b/ao486/pipeline/fetch.v
--- a/ao486/pipeline/fetch.v
+++ b/ao486/pipeline/fetch.v
@@ -1,104 +1,54 @@
-/*
- * Copyright (c) 2014, Aleksander Osman
- * All rights reserved.
- * 
- * Redistribution and use in source and binary forms, with or without
- * modification, are permitted provided that the following conditions are met:
- * 
- * * Redistributions of source code must retain the above copyright notice, this
- *   list of conditions and the following disclaimer.
- * 
- * * Redistributions in binary form must reproduce the above copyright notice,
- *   this list of conditions and the following disclaimer in the documentation
- *   and/or other materials provided with the distribution.
- * 
- * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
- * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
- * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
- * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
- * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
- * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
- * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
- * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
- * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
- * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
- */
-
-`include "defines.v"
-
 module fetch(
-    input               clk,
-    input               rst_n,
-    
-    input               pr_reset,
-    
-    // get prefetch_eip
-    input       [31:0]  wr_eip,
-    
-    output      [31:0]  prefetch_eip,
-    
-    // prefetch_fifo
-    output              prefetchfifo_accept_do,
-    input       [67:0]  prefetchfifo_accept_data,
-    input               prefetchfifo_accept_empty,
-    
-    // fetch interface to decode
-    output      [3:0]   fetch_valid,
-    output      [63:0]  fetch,
-    output              fetch_limit,
-    output              fetch_page_fault,
-    
-    // feedback from decode
-    input       [3:0]   dec_acceptable
+  input         clk,
+                rst_n,
+                pr_reset,
+  input  [31:0] wr_eip,
+  input  [67:0] prefetchfifo_accept_data,
+  input         prefetchfifo_accept_empty,
+  input  [3:0]  dec_acceptable,
+  output [31:0] prefetch_eip,
+  output        prefetchfifo_accept_do,
+  output [3:0]  fetch_valid,
+  output [63:0] fetch,
+  output        fetch_limit,
+                fetch_page_fault
 );
 
-//------------------------------------------------------------------------------
-
-wire partial;
-
-reg [3:0] fetch_count;
-
-//------------------------------------------------------------------------------
-
-assign prefetch_eip = wr_eip;
-
-//------------------------------------------------------------------------------
-
-assign fetch_valid      = (prefetchfifo_accept_empty || prefetchfifo_accept_data[67:64] >= `PREFETCH_MIN_FAULT)? 4'd0 : prefetchfifo_accept_data[67:64] - fetch_count;
-
-assign fetch_limit      = prefetchfifo_accept_empty == `FALSE && prefetchfifo_accept_data[67:64] == `PREFETCH_GP_FAULT;
-assign fetch_page_fault = prefetchfifo_accept_empty == `FALSE && prefetchfifo_accept_data[67:64] == `PREFETCH_PF_FAULT;
-
-assign fetch =
-    (prefetchfifo_accept_empty)?      64'd0 :
-    (fetch_count == 4'd0)?                   prefetchfifo_accept_data[63:0] :
-    (fetch_count == 4'd1)?          {  8'd0, prefetchfifo_accept_data[63:8] } :
-    (fetch_count == 4'd2)?          { 16'd0, prefetchfifo_accept_data[63:16] } :
-    (fetch_count == 4'd3)?          { 24'd0, prefetchfifo_accept_data[63:24] } :
-    (fetch_count == 4'd4)?          { 32'd0, prefetchfifo_accept_data[63:32] } :
-    (fetch_count == 4'd5)?          { 40'd0, prefetchfifo_accept_data[63:40] } :
-    (fetch_count == 4'd6)?          { 48'd0, prefetchfifo_accept_data[63:48] } :
-                                    { 56'd0, prefetchfifo_accept_data[63:56] };
-
-//------------------------------------------------------------------------------
-
-assign prefetchfifo_accept_do   = dec_acceptable >= fetch_valid && prefetchfifo_accept_empty == `FALSE && prefetchfifo_accept_data[67:64] < `PREFETCH_MIN_FAULT;
-
-assign partial                  = dec_acceptable <  fetch_valid && prefetchfifo_accept_empty == `FALSE && prefetchfifo_accept_data[67:64] < `PREFETCH_MIN_FAULT;
-
-//------------------------------------------------------------------------------
-
-always @(posedge clk) begin
-    if(rst_n == 1'b0)               fetch_count <= 4'd0;
-    else if(pr_reset)               fetch_count <= 4'd0;
-    else if(prefetchfifo_accept_do) fetch_count <= 4'd0;
-    else if(partial)                fetch_count <= fetch_count + dec_acceptable;
-end
-
-//------------------------------------------------------------------------------
-                                        
-//------------------------------------------------------------------------------
-
-//------------------------------------------------------------------------------
-
+  reg  [3:0] rt_tmp_1_4;
+  wire       _GEN = ~prefetchfifo_accept_empty & prefetchfifo_accept_data[67:64] < 4'h9;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:10885:17, :10886:17, :10887:17, :10888:17, :10889:18
+  wire [3:0] _GEN_0 =
+    _GEN & rt_tmp_1_4 < prefetchfifo_accept_data[67:64]
+      ? prefetchfifo_accept_data[67:64] - rt_tmp_1_4
+      : 4'h0;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:10884:17, :10886:17, :10889:18, :10890:18, :10891:18, :10893:18, :10905:17
+  wire       _GEN_1 = _GEN & (|_GEN_0);	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:10889:18, :10893:18, :10895:18, :10896:18
+  wire       _GEN_2 = dec_acceptable >= _GEN_0 & _GEN_1;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:10893:18, :10894:18, :10896:18, :10897:18
+  always_ff @(posedge clk)
+    rt_tmp_1_4 <=
+      ~rst_n | pr_reset | _GEN_2
+        ? 4'h0
+        : dec_acceptable < _GEN_0 & _GEN_1 ? rt_tmp_1_4 + dec_acceptable : rt_tmp_1_4;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:10883:17, :10884:17, :10893:18, :10896:18, :10897:18, :10898:18, :10899:18, :10900:18, :10901:18, :10903:18, :10904:18, :10905:17
+  assign prefetch_eip = wr_eip;
+  assign prefetchfifo_accept_do = _GEN_2;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:10897:18, :10957:3
+  assign fetch_valid = _GEN_0;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:10893:18, :10957:3
+  assign fetch =
+    prefetchfifo_accept_empty
+      ? 64'h0
+      : rt_tmp_1_4 == 4'h0
+          ? prefetchfifo_accept_data[63:0]
+          : rt_tmp_1_4 == 4'h1
+              ? {8'h0, prefetchfifo_accept_data[63:8]}
+              : rt_tmp_1_4 == 4'h2
+                  ? {16'h0, prefetchfifo_accept_data[63:16]}
+                  : rt_tmp_1_4 == 4'h3
+                      ? {24'h0, prefetchfifo_accept_data[63:24]}
+                      : rt_tmp_1_4 == 4'h4
+                          ? {32'h0, prefetchfifo_accept_data[63:32]}
+                          : rt_tmp_1_4 == 4'h5
+                              ? {40'h0, prefetchfifo_accept_data[63:40]}
+                              : rt_tmp_1_4 == 4'h6
+                                  ? {48'h0, prefetchfifo_accept_data[63:48]}
+                                  : {56'h0, prefetchfifo_accept_data[63:56]};	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:10884:17, :10905:17, :10906:19, :10908:18, :10909:19, :10910:18, :10911:18, :10912:18, :10913:19, :10914:19, :10915:18, :10916:18, :10917:19, :10918:19, :10919:19, :10920:18, :10921:18, :10922:19, :10923:19, :10924:19, :10925:18, :10926:18, :10927:19, :10928:19, :10929:19, :10930:18, :10931:18, :10932:19, :10933:19, :10934:19, :10935:18, :10936:18, :10937:19, :10938:19, :10939:19, :10940:19, :10941:18, :10942:19, :10943:19, :10944:19, :10945:19, :10946:19, :10947:19, :10948:19, :10949:19, :10950:19, :10957:3
+  assign fetch_limit = ~prefetchfifo_accept_empty & (&(prefetchfifo_accept_data[67:64]));	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:10885:17, :10886:17, :10952:18, :10953:18, :10957:3
+  assign fetch_page_fault =
+    ~prefetchfifo_accept_empty & prefetchfifo_accept_data[67:64] == 4'hE;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:10885:17, :10886:17, :10954:18, :10955:18, :10956:18, :10957:3
 endmodule

