diff --git a/ao486/memory/icache.v b/ao486/memory/icache.v
--- a/ao486/memory/icache.v
+++ b/ao486/memory/icache.v
@@ -1,224 +1,217 @@
-/*
- * Copyright (c) 2014, Aleksander Osman
- * All rights reserved.
- * 
- * Redistribution and use in source and binary forms, with or without
- * modification, are permitted provided that the following conditions are met:
- * 
- * * Redistributions of source code must retain the above copyright notice, this
- *   list of conditions and the following disclaimer.
- * 
- * * Redistributions in binary form must reproduce the above copyright notice,
- *   this list of conditions and the following disclaimer in the documentation
- *   and/or other materials provided with the distribution.
- * 
- * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
- * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
- * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
- * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
- * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
- * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
- * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
- * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
- * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
- * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
- */
-
-`include "defines.v"
-
 module icache(
-    input           clk,
-    input           rst_n,
-    
-    input           cache_disable,
-
-    //RESP:
-    input           pr_reset,
-    
-    input [31:0]    prefetch_address,
-    input [31:0]    delivered_eip,
-    output reg      reset_prefetch = 1'd0,
-    //END
-    
-    //RESP:
-    input           icacheread_do,
-    input   [31:0]  icacheread_address,
-    input   [4:0]   icacheread_length, // takes into account: page size and cs segment limit
-    //END
-    
-    //REQ:
-    output              readcode_do,
-    input               readcode_done,
-    
-    output      [31:0]  readcode_address,
-    input       [31:0]  readcode_partial,
-    //END
-    
-    //REQ:
-    output              prefetchfifo_write_do,
-    output  [35:0]      prefetchfifo_write_data,
-    //END
-    
-    //REQ:
-    output              prefetched_do,
-    output [4:0]        prefetched_length,
-    //END
-    
-    input   [27:2]      snoop_addr,
-    input   [31:0]      snoop_data,
-    input    [3:0]      snoop_be,
-    input               snoop_we
+  input         clk,
+                rst_n,
+                cache_disable,
+                pr_reset,
+  input  [31:0] prefetch_address,
+                delivered_eip,
+  input         icacheread_do,
+  input  [31:0] icacheread_address,
+  input  [4:0]  icacheread_length,
+  input         readcode_done,
+  input  [31:0] readcode_partial,
+  input  [25:0] snoop_addr,
+  input  [31:0] snoop_data,
+  input  [3:0]  snoop_be,
+  input         snoop_we,
+  output        reset_prefetch,
+                readcode_do,
+  output [31:0] readcode_address,
+  output        prefetchfifo_write_do,
+  output [35:0] prefetchfifo_write_data,
+  output        prefetched_do,
+  output [4:0]  prefetched_length
 );
 
-//------------------------------------------------------------------------------
-
-localparam STATE_IDLE = 1'd0;
-localparam STATE_READ = 1'd1;
-
-reg          state;
-reg [4:0]    length;
-reg [11:0]   partial_length;
-reg          reset_waiting;
-             
-wire [4:0]   partial_length_current;
-
-wire [11:0]  length_burst;
-
-wire         readcode_cache_do;
-wire [31:0]  readcode_cache_address;
-wire         readcode_cache_valid;
-wire         readcode_cache_done;
-wire [31:0]  readcode_cache_data;
-
-reg          prefetch_checknext;
-reg  [31:0]  prefetch_checkaddr;
-reg  [31:0]  min_check;
-reg  [31:0]  max_check;
-reg   [1:0]  reset_prefetch_count = 2'd0;
-
-
-//------------------------------------------------------------------------------
-
-wire reset_combined = reset_prefetch | pr_reset;
-
-always @(posedge clk) begin
-   prefetch_checknext <= 1'b0;
-   prefetch_checkaddr <= { 4'd0, snoop_addr, 2'd0 };
-   min_check          <= delivered_eip;
-   max_check          <= prefetch_address + 5'd20; // cache read burst is 16 bytes, so we need to look a bit further, additional + 4 because of 1 cycle delay.
-   
-   if (snoop_we) prefetch_checknext <= 1'b1;
-   
-   if (prefetch_checknext && prefetch_checkaddr >= min_check && prefetch_checkaddr <= max_check) begin
-      reset_prefetch       <= 1'b1;
-      reset_prefetch_count <= 2'd2;
-   end
-   
-   if (reset_prefetch_count > 2'd0) begin
-      reset_prefetch_count <= reset_prefetch_count - 1'd1;
-      if (reset_prefetch_count == 2'd1) reset_prefetch <= 1'd0;
-   end
-   
-end
-
-//------------------------------------------------------------------------------
-
-//MIN(partial_length, length_saved)
-assign partial_length_current =
-    ({ 2'b0, partial_length[2:0] } > length)? length : { 2'b0, partial_length[2:0] };
-    
-//------------------------------------------------------------------------------
-
-always @(posedge clk) begin
-    if(rst_n == 1'b0)                                 reset_waiting <= `FALSE;
-    else if(reset_combined && state != STATE_IDLE)    reset_waiting <= `TRUE;
-    else if(state == STATE_IDLE)                      reset_waiting <= `FALSE;
-end
-
-//------------------------------------------------------------------------------
-
-assign length_burst =
-    (icacheread_address[1:0] == 2'd0)? { 3'd4, 3'd4, 3'd4, 3'd4 } :
-    (icacheread_address[1:0] == 2'd1)? { 3'd4, 3'd4, 3'd4, 3'd3 } :
-    (icacheread_address[1:0] == 2'd2)? { 3'd4, 3'd4, 3'd4, 3'd2 } :
-                                       { 3'd4, 3'd4, 3'd4, 3'd1 };
-                            
-assign prefetchfifo_write_data =
-    (partial_length[2:0] == 3'd1)?   {                  4'd1 ,              24'd0, readcode_cache_data[31:24] } :
-    (partial_length[2:0] == 3'd2)?   { (length > 5'd2)? 4'd2 : length[3:0], 16'd0, readcode_cache_data[31:16] } :
-    (partial_length[2:0] == 3'd3)?   { (length > 5'd3)? 4'd3 : length[3:0],  8'd0, readcode_cache_data[31:8] } :
-                                     { (length > 5'd4)? 4'd4 : length[3:0],        readcode_cache_data[31:0] };
-
-//------------------------------------------------------------------------------
-
-l1_icache l1_icache_inst(
-   
-    .CLK             (clk),
-    .RESET           (~rst_n),
-    .pr_reset        (reset_combined),
-	 
-    .DISABLE         (cache_disable),
-
-    .CPU_REQ         (readcode_cache_do),
-    .CPU_ADDR        (readcode_cache_address),
-    .CPU_VALID       (readcode_cache_valid),
-    .CPU_DONE        (readcode_cache_done),
-    .CPU_DATA        (readcode_cache_data),
-    
-    .MEM_REQ         (readcode_do),
-    .MEM_ADDR        (readcode_address),
-    .MEM_DONE        (readcode_done),
-    .MEM_DATA        (readcode_partial),
-    
-    .snoop_addr      (snoop_addr),
-    .snoop_data      (snoop_data),
-    .snoop_be        (snoop_be),
-    .snoop_we        (snoop_we)
-);
-
-assign readcode_cache_do =
-   (~rst_n) ? (`FALSE) :
-   (state == STATE_IDLE && ~(reset_combined) && icacheread_do && icacheread_length > 5'd0) ? (`TRUE) :
-   `FALSE;
-   
-assign readcode_cache_address = { icacheread_address[31:2], 2'd0 };
-   
-assign prefetchfifo_write_do =
-   (~rst_n) ? (`FALSE) :
-   (state == STATE_READ && reset_combined == `FALSE && reset_waiting == `FALSE && readcode_cache_valid) ? (`TRUE) :
-   `FALSE;
-   
-assign prefetched_length       = partial_length_current;
-
-assign prefetched_do =
-   (~rst_n) ? (`FALSE) :
-   (state == STATE_READ && reset_combined == `FALSE && reset_waiting == `FALSE && readcode_cache_valid) ? (`TRUE) :
-   `FALSE;
-   
-always @(posedge clk) begin
-   if(rst_n == 1'b0) begin
-      state          <= STATE_IDLE;
-      length         <= 5'b0;
-      partial_length <= 12'b0;
-   end
-   else begin
-      if(state == STATE_IDLE && ~(reset_combined) && icacheread_do && icacheread_length > 5'd0) begin
-         state          <= STATE_READ;
-         partial_length <= length_burst;
-         length         <= icacheread_length;
-      end
-      else if (state == STATE_READ) begin
-         if(reset_combined == `FALSE && reset_waiting == `FALSE) begin
-            if(readcode_cache_valid) begin
-               if(partial_length[2:0] > 3'd0 && length > 5'd0) begin
-                  length         <= length - partial_length_current;
-                  partial_length <= { 3'd0, partial_length[11:3] }; 
-               end
-            end
-         end
-         if(readcode_cache_done) state <= STATE_IDLE;
-      end
-   end
-end     
-
+  reg         rt_tmp_20_1;
+  reg  [11:0] rt_tmp_10_12;
+  reg  [4:0]  rt_tmp_9_5;
+  reg         rt_tmp_1_1;
+  reg  [31:0] rt_tmp_2_32;
+  reg  [31:0] rt_tmp_3_32;
+  reg  [31:0] rt_tmp_4_32;
+  reg         rt_tmp_5_1;
+  reg  [1:0]  rt_tmp_6_2;
+  wire        _GEN = rt_tmp_5_1 | pr_reset;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1135:17, :1145:18
+  reg         rt_tmp_7_1;
+  reg         rt_tmp_8_1;
+  wire        _GEN_0 = rt_tmp_20_1 & ~pr_reset;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1184:18, :1185:18, :1372:18
+  wire [4:0]  _GEN_1 = {2'h0, rt_tmp_10_12[2:0]};	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1191:19, :1207:19, :1246:19
+  wire [4:0]  _GEN_2 = _GEN_1 > rt_tmp_9_5 ? rt_tmp_9_5 : _GEN_1;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1207:19, :1208:19, :1209:19, :1215:17
+  reg  [3:0]  rt_tmp_11_4;
+  reg         rt_tmp_12_1;
+  reg         rt_tmp_13_1;
+  reg  [31:0] rt_tmp_14_32;
+  reg         rt_tmp_15_1;
+  reg         rt_tmp_16_1;
+  reg  [3:0]  rt_tmp_17_4;
+  reg         rt_tmp_18_1;
+  reg  [31:0] rt_tmp_19_32;
+  reg  [3:0]  rt_tmp_21_4;
+  reg  [2:0]  rt_tmp_22_3;
+  reg  [31:0] rt_tmp_23_32;
+  reg  [31:0] rt_tmp_24_32;
+  reg  [31:0] rt_tmp_25_32;
+  reg  [31:0] rt_tmp_26_32;
+  reg  [31:0] rt_tmp_27_32;
+  reg  [31:0] rt_tmp_28_32;
+  reg  [31:0] rt_tmp_29_32;
+  reg  [31:0] rt_tmp_30_32;
+  always_ff @(posedge clk) begin
+    automatic logic        _GEN_3 = rt_tmp_6_2 == 2'h0;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1120:18, :1143:17
+    automatic logic        _GEN_4 = rt_tmp_6_2 == 2'h1;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1123:18, :1124:18, :1143:17
+    automatic logic        _GEN_5 =
+      rt_tmp_1_1 & rt_tmp_2_32 >= rt_tmp_3_32 & rt_tmp_2_32 <= rt_tmp_4_32;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1111:17, :1115:18, :1116:18, :1119:18, :1126:18, :1127:18, :1128:18, :1129:18
+    automatic logic        _GEN_6 = rst_n & _GEN & rt_tmp_8_1;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1145:18, :1146:18, :1147:18, :1179:17
+    automatic logic        _GEN_7 =
+      ~rt_tmp_8_1 & ~_GEN & icacheread_do & (|icacheread_length);	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1145:18, :1149:18, :1153:18, :1154:18, :1155:18, :1157:18, :1158:18, :1179:17
+    automatic logic        _GEN_8 = ~_GEN_7 & rst_n & ~rt_tmp_8_1;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1149:18, :1154:18, :1155:18, :1158:18, :1159:18, :1160:18, :1161:18, :1179:17
+    automatic logic        _GEN_9 = rst_n & _GEN_7;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1154:18, :1155:18, :1158:18, :1163:18
+    automatic logic        _GEN_10 = ~rst_n | _GEN_9;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1144:18, :1163:18, :1164:18
+    automatic logic        _GEN_11 = rt_tmp_20_1 & rt_tmp_22_3 == 3'h1;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1165:18, :1166:18, :1167:18, :1372:18, :1394:18
+    automatic logic        _GEN_12 = _GEN_11 | _GEN;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1145:18, :1167:18, :1169:18
+    automatic logic        _GEN_13 = ~_GEN & ~rt_tmp_7_1;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1145:18, :1152:17, :1153:18, :1181:18, :1182:18
+    automatic logic        _GEN_14 = _GEN_13 & ~_GEN_0;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1182:18, :1185:18, :1186:18, :1187:19
+    automatic logic        _GEN_15 =
+      _GEN_0 & _GEN_13 & ~((|(rt_tmp_10_12[2:0])) & (|rt_tmp_9_5));	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1182:18, :1185:18, :1190:19, :1191:19, :1193:19, :1194:19, :1195:19, :1196:19, :1197:19, :1215:17, :1246:19
+    automatic logic        _GEN_16 =
+      ~(~_GEN_8 & (_GEN_10 | _GEN_13 & ~_GEN_14 & ~_GEN_15)) | _GEN_8;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1160:18, :1161:18, :1162:18, :1164:18, :1182:18, :1187:19, :1188:19, :1189:19, :1190:19, :1197:19, :1198:19, :1199:19, :1200:19, :1201:19, :1202:19, :1203:19
+    automatic logic        _GEN_17 = ~_GEN_13 | _GEN_14 | _GEN_15;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1182:18, :1187:19, :1190:19, :1197:19, :1204:19, :1205:19, :1206:19
+    automatic logic [3:0]  _GEN_18 =
+      ~rst_n
+        ? 4'h0
+        : pr_reset
+            ? (rt_tmp_16_1 ? 4'h8 - rt_tmp_17_4 : 4'h0)
+            : (|rt_tmp_11_4) & readcode_done ? rt_tmp_11_4 - 4'h1 : rt_tmp_11_4;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1112:18, :1144:18, :1249:19, :1250:19, :1251:19, :1252:19, :1253:19, :1255:19, :1256:19, :1257:19, :1258:19, :1259:18, :1340:18, :1351:18
+    automatic logic        _GEN_19 = ~rst_n | _GEN;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1144:18, :1145:18, :1261:19
+    automatic logic        _GEN_20 =
+      rst_n & ~rt_tmp_8_1 & ~_GEN & icacheread_do & (|icacheread_length) & ~rt_tmp_20_1;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1145:18, :1149:18, :1153:18, :1157:18, :1179:17, :1264:19, :1267:19, :1268:19, :1270:19, :1271:19, :1272:19, :1372:18
+    automatic logic [31:0] _GEN_21 = {icacheread_address[31:5], 5'h0};
+    automatic logic        _GEN_22 = rt_tmp_18_1 & rt_tmp_19_32 == _GEN_21;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1276:20, :1277:19, :1278:19, :1357:18, :1363:19
+    automatic logic        _GEN_23 = _GEN_20 & _GEN_22;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1264:19, :1267:19, :1268:19, :1270:19, :1272:19, :1278:19, :1279:19
+    automatic logic        _GEN_24 =
+      _GEN_20 & ~_GEN_22 & rt_tmp_16_1 & rt_tmp_14_32 != _GEN_21;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1264:19, :1267:19, :1268:19, :1270:19, :1272:19, :1276:20, :1278:19, :1280:19, :1281:19, :1282:19, :1283:19, :1284:19, :1324:19, :1340:18
+    automatic logic        _GEN_25 = _GEN_20 & ~_GEN_22 & ~rt_tmp_16_1;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1264:19, :1267:19, :1268:19, :1270:19, :1272:19, :1278:19, :1280:19, :1286:19, :1287:19, :1288:19, :1340:18
+    automatic logic        _GEN_26 =
+      rt_tmp_20_1 & (|(rt_tmp_22_3[2:1])) & rt_tmp_21_4 == 4'h7;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1289:19, :1290:19, :1291:19, :1292:19, :1293:19, :1372:18, :1385:18, :1394:18
+    automatic logic        _GEN_27 = _GEN_25 | _GEN_26;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1287:19, :1288:19, :1292:19, :1293:19, :1294:19
+    automatic logic        _GEN_28 =
+      ~_GEN_24 & ~(|rt_tmp_11_4) & readcode_done & (~rt_tmp_12_1 | rt_tmp_13_1);	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1252:19, :1259:18, :1282:19, :1283:19, :1284:19, :1295:19, :1296:19, :1297:19, :1298:19, :1299:19, :1300:19, :1307:18, :1316:18
+    automatic logic        _GEN_29 =
+      ~_GEN_19 & ~_GEN_23
+      & (_GEN_24 | _GEN_27 | (~readcode_done | ~_GEN_28) & rt_tmp_12_1);	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1261:19, :1279:19, :1282:19, :1283:19, :1284:19, :1294:19, :1298:19, :1299:19, :1300:19, :1301:19, :1302:19, :1303:19, :1304:19, :1305:19, :1306:19, :1307:18
+    automatic logic        _GEN_30 = rt_tmp_17_4 == 4'h7;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1290:19, :1326:19, :1351:18
+    automatic logic        _GEN_31 = _GEN_28 & rt_tmp_16_1 & _GEN_30;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1298:19, :1299:19, :1300:19, :1326:19, :1327:19, :1328:19, :1340:18
+    automatic logic        _GEN_32 = rt_tmp_16_1 & _GEN_28;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1298:19, :1299:19, :1300:19, :1340:18, :1341:19
+    automatic logic        _GEN_33 = _GEN_31 & ~rt_tmp_15_1;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1327:19, :1328:19, :1334:18, :1352:19, :1353:19
+    automatic logic        _GEN_34 = _GEN_11 | _GEN_26;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1167:18, :1292:19, :1293:19, :1367:19
+    automatic logic        _GEN_35 = _GEN_23 | _GEN_25 | _GEN_24;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1279:19, :1282:19, :1283:19, :1284:19, :1287:19, :1288:19, :1373:19, :1374:19
+    automatic logic        _GEN_36 = _GEN_32 & ~_GEN_23 & ~rt_tmp_15_1;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1279:19, :1305:19, :1334:18, :1341:19, :1352:19, :1398:19, :1399:19
+    rt_tmp_1_1 <= snoop_we;
+    rt_tmp_2_32 <= {4'h0, snoop_addr, 2'h0};	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1112:18, :1114:19, :1115:18
+    rt_tmp_3_32 <= delivered_eip;
+    rt_tmp_4_32 <= prefetch_address + 32'h14;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1117:19, :1118:19, :1119:18
+    rt_tmp_5_1 <= ~_GEN_3 & _GEN_4 | _GEN_5 ? (_GEN_3 | ~_GEN_4) & _GEN_5 : rt_tmp_5_1;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1120:18, :1122:18, :1124:18, :1125:18, :1127:18, :1129:18, :1130:18, :1131:18, :1132:18, :1133:18, :1134:18, :1135:17
+    rt_tmp_6_2 <=
+      ~_GEN_3 | _GEN_5
+        ? (_GEN_3 ? (_GEN_5 ? 2'h2 : rt_tmp_6_2) : rt_tmp_6_2 - 2'h1)
+        : rt_tmp_6_2;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1120:18, :1122:18, :1127:18, :1129:18, :1136:18, :1137:18, :1138:18, :1140:18, :1141:18, :1142:18, :1143:17
+    rt_tmp_7_1 <= ~rst_n | _GEN_6 | ~rt_tmp_8_1 ? _GEN_6 : rt_tmp_7_1;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1144:18, :1146:18, :1147:18, :1148:18, :1149:18, :1150:18, :1151:18, :1152:17, :1179:17
+    rt_tmp_8_1 <=
+      ~(~_GEN_8 & (_GEN_10 | _GEN_12)) | _GEN_8
+        ? rt_tmp_8_1
+        : rst_n & (_GEN_9 | ~_GEN_12 & rt_tmp_8_1);	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1160:18, :1161:18, :1162:18, :1163:18, :1164:18, :1169:18, :1170:18, :1171:18, :1172:18, :1173:18, :1174:18, :1175:18, :1176:18, :1177:18, :1178:18, :1179:17
+    rt_tmp_9_5 <=
+      _GEN_16
+        ? rt_tmp_9_5
+        : ~rst_n
+            ? 5'h0
+            : _GEN_9 ? icacheread_length : _GEN_17 ? rt_tmp_9_5 : rt_tmp_9_5 - _GEN_2;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1144:18, :1163:18, :1203:19, :1205:19, :1206:19, :1209:19, :1210:19, :1211:19, :1212:19, :1213:19, :1214:19, :1215:17
+    rt_tmp_10_12 <=
+      _GEN_16
+        ? rt_tmp_10_12
+        : ~rst_n
+            ? 12'h0
+            : _GEN_9
+                ? (~(icacheread_address[1])
+                     ? (icacheread_address[1:0] == 2'h0 | icacheread_address[1:0] != 2'h1
+                          ? 12'h924
+                          : 12'h923)
+                     : icacheread_address[1:0] != 2'h3
+                         ? (icacheread_address[1:0] == 2'h2 ? 12'h922 : 12'h924)
+                         : (&(icacheread_address[1:0])) ? 12'h921 : 12'h924)
+                : _GEN_17 ? rt_tmp_10_12 : {3'h0, rt_tmp_10_12[11:3]};	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1123:18, :1137:18, :1139:18, :1144:18, :1163:18, :1192:19, :1203:19, :1205:19, :1206:19, :1216:20, :1217:19, :1219:19, :1221:19, :1222:20, :1224:19, :1225:20, :1227:20, :1229:19, :1231:19, :1232:20, :1233:20, :1235:19, :1236:20, :1237:20, :1238:20, :1239:20, :1240:19, :1241:20, :1242:20, :1243:20, :1244:20, :1245:20, :1246:19
+    rt_tmp_11_4 <= _GEN_18;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1258:19, :1259:18
+    rt_tmp_12_1 <= _GEN_29;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1305:19, :1306:19, :1307:18
+    rt_tmp_13_1 <=
+      ~(_GEN_23 | _GEN_24) & ~_GEN_27 & _GEN_29 & ~((|_GEN_18) | readcode_done);	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1258:19, :1279:19, :1282:19, :1283:19, :1284:19, :1294:19, :1305:19, :1306:19, :1308:19, :1309:19, :1310:19, :1311:19, :1312:19, :1313:19, :1314:19, :1315:19, :1316:18
+    rt_tmp_14_32 <=
+      _GEN_19
+        ? 32'h0
+        : _GEN_25 | _GEN_24 ? _GEN_21 : _GEN_26 ? rt_tmp_19_32 + 32'h20 : rt_tmp_14_32;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1261:19, :1276:20, :1282:19, :1283:19, :1284:19, :1287:19, :1288:19, :1292:19, :1293:19, :1317:20, :1318:20, :1319:20, :1320:20, :1322:20, :1323:20, :1324:19, :1363:19
+    rt_tmp_15_1 <=
+      ~_GEN_19 & ~_GEN_23 & ~_GEN_24
+      & (_GEN_27 ? _GEN_25 & rt_tmp_20_1 : ~_GEN_31 & rt_tmp_15_1);	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1261:19, :1279:19, :1282:19, :1283:19, :1284:19, :1287:19, :1288:19, :1294:19, :1295:19, :1305:19, :1306:19, :1325:19, :1327:19, :1328:19, :1329:19, :1330:19, :1331:19, :1332:19, :1333:19, :1334:18, :1372:18
+    rt_tmp_16_1 <= ~_GEN_19 & ~_GEN_23 & (_GEN_24 | _GEN_27 | ~_GEN_31 & rt_tmp_16_1);	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1261:19, :1279:19, :1282:19, :1283:19, :1284:19, :1294:19, :1305:19, :1306:19, :1327:19, :1328:19, :1329:19, :1335:19, :1336:19, :1337:19, :1338:19, :1339:19, :1340:18
+    rt_tmp_17_4 <=
+      _GEN_19 | _GEN_23 | _GEN_24 | _GEN_27
+        ? 4'h0
+        : _GEN_32 ? (_GEN_30 ? 4'h0 : rt_tmp_17_4 + 4'h1) : rt_tmp_17_4;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1112:18, :1254:19, :1261:19, :1279:19, :1282:19, :1283:19, :1284:19, :1294:19, :1326:19, :1341:19, :1344:19, :1345:19, :1346:19, :1348:19, :1349:19, :1350:19, :1351:18
+    rt_tmp_18_1 <= ~_GEN_19 & (~_GEN_23 & _GEN_33 | rt_tmp_18_1);	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1261:19, :1279:19, :1305:19, :1306:19, :1353:19, :1355:19, :1356:19, :1357:18
+    rt_tmp_19_32 <= _GEN_19 ? 32'h0 : _GEN_23 | ~_GEN_33 ? rt_tmp_19_32 : rt_tmp_14_32;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1261:19, :1279:19, :1317:20, :1324:19, :1353:19, :1361:20, :1362:20, :1363:19
+    rt_tmp_20_1 <=
+      ~_GEN_19 & (_GEN_23 | _GEN_33 | (rt_tmp_20_1 ? ~_GEN_34 : rt_tmp_20_1));	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1261:19, :1279:19, :1306:19, :1353:19, :1366:19, :1367:19, :1368:19, :1369:19, :1370:19, :1371:19, :1372:18
+    rt_tmp_21_4 <=
+      _GEN_19
+        ? 4'h0
+        : _GEN_35
+            ? {1'h0, icacheread_address[4:2]}
+            : _GEN_31 | ~rt_tmp_20_1 ? rt_tmp_21_4 : _GEN_34 ? 4'h0 : rt_tmp_21_4 + 4'h1;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1112:18, :1254:19, :1261:19, :1262:19, :1271:19, :1327:19, :1328:19, :1367:19, :1372:18, :1373:19, :1374:19, :1375:19, :1376:19, :1378:19, :1380:19, :1382:19, :1383:19, :1384:19, :1385:18
+    rt_tmp_22_3 <=
+      _GEN_19 ? 3'h0 : _GEN_35 ? 3'h4 : rt_tmp_20_1 ? rt_tmp_22_3 - 3'h1 : rt_tmp_22_3;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1192:19, :1261:19, :1372:18, :1373:19, :1374:19, :1389:19, :1390:19, :1391:19, :1392:19, :1393:19, :1394:18
+    rt_tmp_23_32 <=
+      _GEN_19 ? 32'h0 : _GEN_36 & rt_tmp_17_4 == 4'h0 ? readcode_partial : rt_tmp_23_32;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1112:18, :1261:19, :1317:20, :1351:18, :1398:19, :1399:19, :1401:19, :1402:19, :1403:20, :1404:20, :1405:19
+    rt_tmp_24_32 <=
+      _GEN_19 ? 32'h0 : _GEN_36 & rt_tmp_17_4 == 4'h1 ? readcode_partial : rt_tmp_24_32;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1254:19, :1261:19, :1317:20, :1351:18, :1398:19, :1399:19, :1407:19, :1408:19, :1409:20, :1410:20, :1411:19
+    rt_tmp_25_32 <=
+      _GEN_19 ? 32'h0 : _GEN_36 & rt_tmp_17_4 == 4'h2 ? readcode_partial : rt_tmp_25_32;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1261:19, :1317:20, :1351:18, :1398:19, :1399:19, :1412:19, :1413:19, :1414:19, :1415:20, :1416:20, :1417:19
+    rt_tmp_26_32 <=
+      _GEN_19 ? 32'h0 : _GEN_36 & rt_tmp_17_4 == 4'h3 ? readcode_partial : rt_tmp_26_32;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1261:19, :1317:20, :1351:18, :1398:19, :1399:19, :1418:19, :1419:19, :1420:19, :1421:20, :1422:20, :1423:19
+    rt_tmp_27_32 <=
+      _GEN_19 ? 32'h0 : _GEN_36 & rt_tmp_17_4 == 4'h4 ? readcode_partial : rt_tmp_27_32;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1261:19, :1317:20, :1351:18, :1398:19, :1399:19, :1424:19, :1425:19, :1426:19, :1427:20, :1428:20, :1429:19
+    rt_tmp_28_32 <=
+      _GEN_19 ? 32'h0 : _GEN_36 & rt_tmp_17_4 == 4'h5 ? readcode_partial : rt_tmp_28_32;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1261:19, :1317:20, :1351:18, :1398:19, :1399:19, :1430:19, :1431:19, :1432:19, :1433:20, :1434:20, :1435:19
+    rt_tmp_29_32 <=
+      _GEN_19 ? 32'h0 : _GEN_36 & rt_tmp_17_4 == 4'h6 ? readcode_partial : rt_tmp_29_32;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1261:19, :1317:20, :1351:18, :1398:19, :1399:19, :1436:19, :1437:19, :1438:19, :1439:20, :1440:20, :1441:19
+    rt_tmp_30_32 <= _GEN_19 ? 32'h0 : _GEN_36 & _GEN_30 ? readcode_partial : rt_tmp_30_32;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1261:19, :1317:20, :1326:19, :1398:19, :1399:19, :1444:19, :1445:20, :1446:20, :1447:19
+  end // always_ff @(posedge)
+  wire        _GEN_37 = rst_n & rt_tmp_8_1 & ~_GEN & ~rt_tmp_7_1 & _GEN_0;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1145:18, :1152:17, :1153:18, :1179:17, :1181:18, :1185:18, :1448:19, :1449:19, :1451:19, :1452:19
+  wire [31:0] _GEN_38 =
+    rt_tmp_21_4 == 4'h0
+      ? rt_tmp_23_32
+      : rt_tmp_21_4 == 4'h1
+          ? rt_tmp_24_32
+          : rt_tmp_21_4 == 4'h2
+              ? rt_tmp_25_32
+              : rt_tmp_21_4 == 4'h3
+                  ? rt_tmp_26_32
+                  : rt_tmp_21_4 == 4'h4
+                      ? rt_tmp_27_32
+                      : rt_tmp_21_4 == 4'h5
+                          ? rt_tmp_28_32
+                          : rt_tmp_21_4 == 4'h6 ? rt_tmp_29_32 : rt_tmp_30_32;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1112:18, :1254:19, :1385:18, :1405:19, :1411:19, :1412:19, :1417:19, :1418:19, :1423:19, :1424:19, :1429:19, :1430:19, :1435:19, :1436:19, :1441:19, :1447:19, :1458:19, :1460:19, :1462:19, :1464:19, :1466:19, :1468:19, :1470:19, :1474:20, :1475:20, :1476:20, :1477:20, :1478:20, :1479:20, :1480:20
+  assign reset_prefetch = rt_tmp_5_1;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1135:17, :1513:3
+  assign readcode_do = rt_tmp_12_1;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1307:18, :1513:3
+  assign readcode_address = rt_tmp_14_32;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1324:19, :1513:3
+  assign prefetchfifo_write_do = _GEN_37;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1448:19, :1449:19, :1451:19, :1452:19, :1513:3
+  assign prefetchfifo_write_data =
+    rt_tmp_10_12[2:0] == 3'h1
+      ? {28'h1000000, _GEN_38[31:24]}
+      : rt_tmp_10_12[2:0] == 3'h2
+          ? {rt_tmp_9_5 > 5'h2 ? 4'h2 : rt_tmp_9_5[3:0], 16'h0, _GEN_38[31:16]}
+          : rt_tmp_10_12[2:0] == 3'h3
+              ? {(|(rt_tmp_9_5[4:2])) ? 4'h3 : rt_tmp_9_5[3:0], 8'h0, _GEN_38[31:8]}
+              : {rt_tmp_9_5 > 5'h4 ? 4'h4 : rt_tmp_9_5[3:0], _GEN_38};	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1165:18, :1191:19, :1215:17, :1246:19, :1412:19, :1418:19, :1424:19, :1455:19, :1456:20, :1480:20, :1481:19, :1482:20, :1483:19, :1484:19, :1485:19, :1486:19, :1488:19, :1489:19, :1490:20, :1491:20, :1492:20, :1493:19, :1494:19, :1495:19, :1496:19, :1498:19, :1499:19, :1500:20, :1501:20, :1502:19, :1503:19, :1505:19, :1506:20, :1507:20, :1508:20, :1509:20, :1513:3
+  assign prefetched_do = _GEN_37;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1448:19, :1449:19, :1451:19, :1452:19, :1513:3
+  assign prefetched_length = _GEN_2;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/runner.mlir:1209:19, :1513:3
 endmodule

