diff --git a/T1-CPU/ifu/sparc_ifu_swl.v b/T1-CPU/ifu/sparc_ifu_swl.v
--- a/T1-CPU/ifu/sparc_ifu_swl.v
+++ b/T1-CPU/ifu/sparc_ifu_swl.v
@@ -196,7 +196,7 @@
 // Declarations
 //----------------------------------------------------------
    // local signals
-   wire [7:0]    count_nxt,
+   wire [9:0]    count_nxt,
 	               count;
    wire          proc0;
    wire          start_on_rst;
@@ -526,18 +526,18 @@
    //---------------------------------------------
    // Start off thread on reset using this counter
    //---------------------------------------------
-   dffr_s #(8) thrrdy_ctr(.din (count_nxt),
+   dffr_s #(10) thrrdy_ctr(.din (count_nxt),
 		                    .clk (clk),
 		                    .q   (count),
 		                    .rst (dtu_reset),
 		                    .se (se), .si(), .so());
 
-   // count_nxt = count + 1, sticky at 128 = 11111111
-   assign count_nxt[7:0] = count[7] ? 8'hFF :
-                           (count[7:0] + 8'h01);
+   // count_nxt = count + 1, sticky at 512 = 10'b11_1111_1111
+   assign count_nxt[9:0] = count[9] ? 10'h3FF :
+                           (count[9:0] + 10'h001);
 
    assign proc0 = (const_cpuid == 4'b0000) ? 1'b1 : 1'b0;
-   assign start_on_rst = (~count[7]) & proc0;
+   assign start_on_rst = (~count[9]) & proc0;
 
 //`ifdef IFU_SAT   
 //   // temporary hack to start threads
diff --git a/T1-CPU/lsu/lsu_qctl2.v b/T1-CPU/lsu/lsu_qctl2.v
--- a/T1-CPU/lsu/lsu_qctl2.v
+++ b/T1-CPU/lsu/lsu_qctl2.v
@@ -492,8 +492,8 @@
 wire          cpx_fwd_req_ic ;
 wire          dfq_fwd_req_ic_type ;
 wire          dfq_rd_vld_d1 ;
-wire  [7:0]   fast_boot_count;
-wire  [7:0]   fast_boot_count_nxt;
+wire  [9:0]   fast_boot_count;
+wire  [9:0]   fast_boot_count_nxt;
 wire          fast_boot_proc0;
 wire          fast_boot_stall_bypass;
 
@@ -506,16 +506,16 @@
 assign  reset  =  ~dbb_reset_l;
 assign  clk = rclk;
 
-dffr_s #(8) fast_boot_ctr(.din (fast_boot_count_nxt),
+dffr_s #(10) fast_boot_ctr(.din (fast_boot_count_nxt),
                           .clk (clk),
                           .q   (fast_boot_count),
                           .rst (reset),
                           .se  (se), .si(), .so());
 
-assign fast_boot_count_nxt[7:0] = fast_boot_count[7] ? 8'hFF :
-                                  (fast_boot_count[7:0] + 8'h01);
+assign fast_boot_count_nxt[9:0] = fast_boot_count[9] ? 10'h3FF :
+                                  (fast_boot_count[9:0] + 10'h001);
 assign fast_boot_proc0 = (const_cpuid[2:0] == 3'b000) ? 1'b1 : 1'b0;
-assign fast_boot_stall_bypass = fast_boot_proc0 & ~fast_boot_count[7];
+assign fast_boot_stall_bypass = fast_boot_proc0 & ~fast_boot_count[9];
 
 
 
