diff --git a/T1-CPU/lsu/lsu_qctl1.v b/T1-CPU/lsu/lsu_qctl1.v
--- a/T1-CPU/lsu/lsu_qctl1.v
+++ b/T1-CPU/lsu/lsu_qctl1.v
@@ -1007,7 +1007,7 @@
 	        .se     (1'b0),       .si (),          .so ()
 	        ); */
 
-assign	lsu_ifu_pcxpkt_ack_d = imiss_pcx_rq_sel_d2 & ~pcx_req_squash_d1 ;
+assign	lsu_ifu_pcxpkt_ack_d = ifu_lsu_pcxreq_d & ~pcx_req_squash_d1 ;
 
 assign  imiss_pkt_vld =  ifu_lsu_pcxreq_d & ~(imiss_pcx_rq_sel_d1 | imiss_pcx_rq_sel_d2) ;
