diff --git a/os2wb/os2wb_dual.v b/os2wb/os2wb_dual.v
--- a/os2wb/os2wb_dual.v
+++ b/os2wb/os2wb_dual.v
@@ -329,14 +329,9 @@ always @(posedge clk or negedge rstn)
                end               
          `WAKEUP:
             begin
-               if(ready)
-                  begin
-                     cpx_packet<=145'h1700000000000000000000000000000010001;
-                     cpx_ready<=1;
-                     state<=`PCX_IDLE;
-                  end
-               else
-                  cpx_ready<=0;
+               cpx_packet<=145'b0;
+               cpx_ready<=0;
+               state<=`PCX_IDLE;
             end
          `PCX_IDLE:
             begin
