diff --git a/ao486/memory/prefetch_fifo.v b/ao486/memory/prefetch_fifo.v
--- a/ao486/memory/prefetch_fifo.v
+++ b/ao486/memory/prefetch_fifo.v
@@ -1,101 +1,82 @@
-/*
- * Copyright (c) 2014, Aleksander Osman
- * All rights reserved.
- * 
- * Redistribution and use in source and binary forms, with or without
- * modification, are permitted provided that the following conditions are met:
- * 
- * * Redistributions of source code must retain the above copyright notice, this
- *   list of conditions and the following disclaimer.
- * 
- * * Redistributions in binary form must reproduce the above copyright notice,
- *   this list of conditions and the following disclaimer in the documentation
- *   and/or other materials provided with the distribution.
- * 
- * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
- * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
- * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
- * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
- * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
- * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
- * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
- * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
- * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
- * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
- */
-
-`include "defines.v"
-
 module prefetch_fifo(
-    input           clk,
-    input           rst_n,
-    
-    input           pr_reset,
-    
-    //RESP:
-    input           prefetchfifo_signal_limit_do,
-    //END
-    
-    //RESP:
-    input           prefetchfifo_signal_pf_do,
-    //END
-    
-    //RESP:
-    input           prefetchfifo_write_do,
-    input   [35:0]  prefetchfifo_write_data,
-    //END
-    
-    output  [4:0]   prefetchfifo_used,
-    
-    //RESP:
-    input           prefetchfifo_accept_do,
-    output [67:0]   prefetchfifo_accept_data,
-    output          prefetchfifo_accept_empty
-    //END
+  input         clk,
+                rst_n,
+                pr_reset,
+                prefetchfifo_signal_limit_do,
+                prefetchfifo_signal_pf_do,
+                prefetchfifo_write_do,
+  input  [35:0] prefetchfifo_write_data,
+  input         prefetchfifo_accept_do,
+  output [4:0]  prefetchfifo_used,
+  output [67:0] prefetchfifo_accept_data,
+  output        prefetchfifo_accept_empty
 );
 
-//------------------------------------------------------------------------------
-
-wire [35:0] q;
-wire        empty;
-wire        bypass;
-
-assign bypass = prefetchfifo_write_do && empty;
-
-assign prefetchfifo_accept_data = (bypass) ? { prefetchfifo_write_data[35:32], 32'd0, prefetchfifo_write_data[31:0] } : { q[35:32], 32'd0, q[31:0] };
-
-assign prefetchfifo_accept_empty = empty && ~bypass;
-
-//------------------------------------------------------------------------------
-
-simple_fifo_mlab #(
-    .width      (36),
-    .widthu     (4)
-)
-prefetch_fifo_inst(
-    .clk        (clk),      //input
-    .rst_n      (rst_n),    //input
-    .sclr       (pr_reset), //input
-    
-    .rdreq      (prefetchfifo_accept_do),                                                               //input
-    .wrreq      ((prefetchfifo_write_do && (~empty || ~prefetchfifo_accept_do)) || prefetchfifo_signal_limit_do || prefetchfifo_signal_pf_do),   //input
-    .data       ((prefetchfifo_signal_limit_do)? { `PREFETCH_GP_FAULT, 32'd0 } :
-                 (prefetchfifo_signal_pf_do)?    { `PREFETCH_PF_FAULT, 32'd0 } :
-                                                     prefetchfifo_write_data),                          //input [35:0]
-    
-    
-    .empty      (empty),                    //output
-    .full       (prefetchfifo_used[4]),     //output
-    .q          (q),                        //output [35:0]
-    .usedw      (prefetchfifo_used[3:0])    //output [3:0]
-);
-
-//------------------------------------------------------------------------------
-
-//------------------------------------------------------------------------------
-
-//------------------------------------------------------------------------------
-
-//------------------------------------------------------------------------------
-
+  reg  [4:0]  rt_tmp_1_5;
+  wire        _GEN = rt_tmp_1_5 == 5'h0;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:2024:18, :2025:18, :2045:17
+  reg  [35:0] rt_tmp_2_36;
+  reg  [35:0] rt_tmp_3_36;
+  reg  [35:0] rt_tmp_4_36;
+  reg  [35:0] rt_tmp_5_36;
+  reg  [35:0] rt_tmp_6_36;
+  reg  [35:0] rt_tmp_7_36;
+  reg  [35:0] rt_tmp_8_36;
+  reg  [35:0] rt_tmp_9_36;
+  always_ff @(posedge clk) begin
+    automatic logic        _GEN_0 = ~rst_n | pr_reset;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:2022:18, :2023:18
+    automatic logic        _GEN_1 = prefetchfifo_accept_do & ~_GEN;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:2025:18, :2026:18, :2027:18
+    automatic logic        _GEN_2 =
+      (prefetchfifo_write_do & (~_GEN | ~prefetchfifo_accept_do)
+       | prefetchfifo_signal_limit_do | prefetchfifo_signal_pf_do)
+      & (rt_tmp_1_5[4:3] == 2'h0 | _GEN_1);	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:2025:18, :2026:18, :2027:18, :2028:18, :2029:18, :2030:18, :2031:18, :2032:18, :2034:18, :2036:18, :2037:18, :2045:17
+    automatic logic [4:0]  _GEN_3 = rt_tmp_1_5 - 5'h1;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:2039:18, :2045:17
+    automatic logic [4:0]  _GEN_4 = _GEN_1 ? _GEN_3 : rt_tmp_1_5;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:2027:18, :2039:18, :2045:17, :2047:18
+    automatic logic [35:0] _GEN_5 =
+      prefetchfifo_signal_limit_do
+        ? 36'hF00000000
+        : prefetchfifo_signal_pf_do ? 36'hE00000000 : prefetchfifo_write_data;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:2053:19, :2055:19, :2056:19, :2057:19
+    rt_tmp_1_5 <=
+      _GEN_0
+        ? 5'h0
+        : _GEN_1
+            ? (_GEN_2 ? rt_tmp_1_5 : _GEN_3)
+            : _GEN_2 ? rt_tmp_1_5 + 5'h1 : rt_tmp_1_5;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:2023:18, :2024:18, :2027:18, :2037:18, :2038:18, :2039:18, :2040:18, :2041:18, :2042:18, :2043:18, :2044:18, :2045:17
+    rt_tmp_2_36 <=
+      _GEN_0
+        ? 36'h0
+        : _GEN_2 & _GEN_4 == 5'h0 ? _GEN_5 : _GEN_1 ? rt_tmp_3_36 : rt_tmp_2_36;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:2023:18, :2024:18, :2027:18, :2037:18, :2046:19, :2047:18, :2049:18, :2050:18, :2057:19, :2058:19, :2059:19, :2060:19, :2061:18, :2068:18
+    rt_tmp_3_36 <=
+      _GEN_0
+        ? 36'h0
+        : _GEN_2 & _GEN_4 == 5'h1 ? _GEN_5 : _GEN_1 ? rt_tmp_4_36 : rt_tmp_3_36;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:2023:18, :2027:18, :2037:18, :2038:18, :2046:19, :2047:18, :2057:19, :2063:18, :2064:18, :2065:19, :2066:19, :2067:19, :2068:18, :2075:18
+    rt_tmp_4_36 <=
+      _GEN_0
+        ? 36'h0
+        : _GEN_2 & _GEN_4 == 5'h2 ? _GEN_5 : _GEN_1 ? rt_tmp_5_36 : rt_tmp_4_36;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:2023:18, :2027:18, :2037:18, :2046:19, :2047:18, :2057:19, :2069:18, :2070:18, :2071:18, :2072:19, :2073:19, :2074:19, :2075:18, :2082:18
+    rt_tmp_5_36 <=
+      _GEN_0
+        ? 36'h0
+        : _GEN_2 & _GEN_4 == 5'h3 ? _GEN_5 : _GEN_1 ? rt_tmp_6_36 : rt_tmp_5_36;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:2023:18, :2027:18, :2037:18, :2046:19, :2047:18, :2057:19, :2076:18, :2077:18, :2078:18, :2079:19, :2080:19, :2081:19, :2082:18, :2089:18
+    rt_tmp_6_36 <=
+      _GEN_0
+        ? 36'h0
+        : _GEN_2 & _GEN_4 == 5'h4 ? _GEN_5 : _GEN_1 ? rt_tmp_7_36 : rt_tmp_6_36;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:2023:18, :2027:18, :2037:18, :2046:19, :2047:18, :2057:19, :2083:18, :2084:18, :2085:18, :2086:19, :2087:19, :2088:19, :2089:18, :2096:18
+    rt_tmp_7_36 <=
+      _GEN_0
+        ? 36'h0
+        : _GEN_2 & _GEN_4 == 5'h5 ? _GEN_5 : _GEN_1 ? rt_tmp_8_36 : rt_tmp_7_36;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:2023:18, :2027:18, :2037:18, :2046:19, :2047:18, :2057:19, :2090:18, :2091:18, :2092:18, :2093:19, :2094:19, :2095:19, :2096:18, :2103:18
+    rt_tmp_8_36 <=
+      _GEN_0
+        ? 36'h0
+        : _GEN_2 & _GEN_4 == 5'h6 ? _GEN_5 : _GEN_1 ? rt_tmp_9_36 : rt_tmp_8_36;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:2023:18, :2027:18, :2037:18, :2046:19, :2047:18, :2057:19, :2097:18, :2098:18, :2099:18, :2100:19, :2101:19, :2102:19, :2103:18, :2110:18
+    rt_tmp_9_36 <=
+      _GEN_0 ? 36'h0 : _GEN_2 & _GEN_4 == 5'h7 ? _GEN_5 : _GEN_1 ? 36'h0 : rt_tmp_9_36;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:2023:18, :2027:18, :2037:18, :2046:19, :2047:18, :2057:19, :2104:18, :2105:18, :2106:18, :2107:19, :2108:19, :2109:19, :2110:18
+  end // always_ff @(posedge)
+  wire        _GEN_6 = prefetchfifo_write_do & _GEN;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:2025:18, :2111:18
+  assign prefetchfifo_used = rt_tmp_1_5;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:2045:17, :2121:3
+  assign prefetchfifo_accept_data =
+    _GEN_6
+      ? {prefetchfifo_write_data[35:32], 32'h0, prefetchfifo_write_data[31:0]}
+      : {rt_tmp_2_36[35:32], 32'h0, rt_tmp_2_36[31:0]};	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:2052:19, :2061:18, :2111:18, :2112:18, :2113:19, :2114:19, :2115:18, :2116:19, :2117:19, :2118:19, :2121:3
+  assign prefetchfifo_accept_empty = _GEN & ~_GEN_6;	// /var/folders/fx/2jqq5cks33g297zlybsqw6c40000gn/T/ao486_patch_profile_ws20260311-55795-e9mvr1/parity.mlir:2025:18, :2111:18, :2119:19, :2120:19, :2121:3
 endmodule

